File: cpp/693_has_alternating_bits/has_alternating_bits.cpp
#include <cassert>
#include <tuple>
#include <vector>

using namespace std;

class Solution {
 public:
  bool hasAlternatingBits(int n) {
    long a = n ^ (n >> 1);
    return (a & (a + 1)) == 0;
  }
};

int main() {
  vector<tuple<int, bool>> tests{
      {5, true},
      {7, false},
      {11, false},
  };

  for (auto& [n, ans] : tests) {
    assert(Solution().hasAlternatingBits(n) == ans);
  }
}

File: rust/693_has_alternating_bits/has_alternating_bits.rs
struct Solution;

impl Solution {
    pub fn has_alternating_bits(mut n: i32) -> bool {
        n ^= n >> 1;
        (n & n + 1) == 0
    }
}

fn main() {
    let tests = vec![(5, true), (7, false), (11, false)];

    for (n, ans) in tests {
        assert_eq!(Solution::has_alternating_bits(n), ans);
    }
}
